### DISCUSSION TITLE
Multiple steps embedded into one URL

### DISCUSSION THREAD

**Discussion by johnlatz** [dvehrs/podget#76] [Q&A]
Starting with the file:  https://feeds.npr.org/510208/podcast.xml

Each show is identified as: 
enclosure length="32582514" type="audio/mpeg" url="https://chrt.fm/track/138C95/prfx.byspotify.com/e/play.podtrac.com/npr-510208/traffic.megaphone.fm/NPR7323383308.mp3?d=2036&size=32582514&e=1196761015&t=podcast&p=510208"/

Podget will contact chrt.fm, and be redirected to contact prfx.byspotify.com - at which point podget will hang...
Resolving chrt.fm (chrt.fm)... 216.137.39.102, 216.137.39.7, 216.137.39.72, ...
Connecting to chrt.fm (chrt.fm)|216.137.39.102|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://prfx.byspotify.com/e/play.podtrac.com/npr-510208/traffic.megaphone.fm/NPR4194073454.mp3?d=2031&size=32506863&e=1196761052&t=podcast&p=510208 [following]
--2024-07-19 12:17:38--  https://prfx.byspotify.com/e/play.podtrac.com/npr-510208/traffic.megaphone.fm/NPR4194073454.mp3?d=2031&size=32506863&e=1196761052&t=podcast&p=510208
Resolving prfx.byspotify.com (prfx.byspotify.com)... 34.110.159.9
Connecting to prfx.byspotify.com (prfx.byspotify.com)|34.110.159.9|:443... failed: Connection timed out.
Retrying.

If I manually attempt to grab the file via wget:
wget "https://traffic.megaphone.fm/NPR4194073454.mp3?d=2031&size=32506863&e=1196761052&t=podcast&p=510208"

it will download.  But that defeats the purpose of an automated script go grab it for me.

and slapping "https://prfx.byspotify.com/e/play.podtrac.com/npr-510208/traffic.megaphone.fm/NPR4194073454.mp3?d=2031&size=32506863&e=1196761052&t=podcast&p=510208" into the URL bar of a browser jumps from spotify to megaphone and starts the file playing just fine.  So I suspect the issue is that Spotify doesn't like podget/wget

Any suggestions?  Is there a means to have podget spoof it's identity as a browser?  Is there a means to have podget simply skip to the last obvious URL in the actual link?

**Comment by dvehrs** [comment]
In our most recent update we attempted to address this issue.

Prior to version 0.9.3 we found that some feeds did not like the default Wget user agent and would not allow us to download.  So we added the ability to set the user agent to a custom value in the WGET_BASEOPTS value in your podgetrc file.  Now from my testing, setting the user agent to "Podget" worked for everything that was blocking me so that is the new default user agent.

Now we also tested with an actual user agent string from a browser and it also worked most of the time but was a bit harder to use because you had to deal with quotes inside a quoted string that was parsed a few times.

Simple examples:  
```
WGET_BASEOPTS="-c -nH --user-agent=Mozilla"
WGET_BASEOPTS="-c -nH --user-agent=Chrome"
WGET_BASEOPTS="-c -nH --user-agent=AppleWebKit"
WGET_BASEOPTS="-c -nH --user-agent=anybrowser"
WGET_BASEOPTS="-c -nH --user-agent=Surprise"
WGET_BASEOPTS="-c -nH --user-agent=Podget/Wget"
```

More Complicated Examples:
```
WGET_BASEOPTS="-c -nH --user-agent=\\\"Dread Pirate Roberts\\\""
WGET_BASEOPTS="-c -nH --user-agent=\\\"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/117.0.0.0 Safari/537.36\\\""
```


Interesting we found that using a user agent of "Wget" without the version number attached would also often work.
```
WGET_BASEOPTS="-c -nH --user-agent=Wget"
```

Now if you are playing with user agents but want one specific feed to use the default Wget user agent then you can set that in your serverlist by adding OPT_WGET_DEFUSERAGENT to its configuration line.  This will take Podget back to the older style for that one feed and the Wget user agent will be something like "Wget/1.21.2".

Now all of this assumes you are using Podget version 0.9.3.   If you aren't then update first then start playing with it.  You may find the new version works with the new defaults.  It all depends on how picky the remote side is going to be as to which user agents it accepts.  

Dave

**Comment by johnlatz** [comment]
Well I can definitively say there is absolutely no issue with the podget code wrt my problems - I've been attempting to send different user agents to Spotify directly from wget, and no joy...

Thanks for your quick response.

**Comment by dvehrs** [comment]
OK, I just ran the feed:  https://feeds.npr.org/510208/podcast.xml

I tested this feed in three versions:  
- 0.9.0  Older version before we allowed for customized user agents.
- 0.9.3  Current release version with default user agent of 'Podget'
- Latest development branch I've been working on (also with same user agent as 0.9.3)

This feed worked just fine in all three versions.  Wget was able to follow the file location changes just fine.

It may be that what you are seeing is because of the delay  in the download that is exceeding Wget's default timeout.  You may experiement with adding '--timeout=90'  with different values in seconds to your WGET_BASEOPTS to see if that is the issue.  

Setting '--timeout=0'  will disable the timeouts function for Wget completely.  That may save you time in trying multiple values until you find the right one. 

Good Luck,
Dave

**Comment by johnlatz** [comment]
Dammit

An overaggressive iptables entry...
